### PR TITLE
Send notifications for mentions in pull request code comments

### LINKED ISSUE #13450: [BUG] Not receiving any emails with setting "Only Email on Mention"
- Gitea version (or commit ref): 1.13 rc1
- Git version: 2.28.0
- Operating system: Ubuntu 18.04 x64, Release downloaded from Gitea.io, Gitea is currently running as a service, service file is very similar to the example service file in docs
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [* ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ *] Yes (https://try.gitea.io/zak/dsadsadsadas/issues/1#issuecomment-69976)
  - [ ] No
- Log gist:
Logs don't show anything really, no failed to send an email or any of that. 

## Description

After setting the DEFAULT_EMAIL_NOTIFICATIONS setting to onmention, and having all current devs update their already existing email setting to onmention. No one is receiving emails now even if they were mentioned. Prior to this everyone who was watching the repo was receiving emails. This wasn't ideal for us so we switched over to mention only, and now we get no emails at all. This seems very similar to #12774 just a different version.  Additionally, if I use the send testing email button, I receive an email. So it seems everything related to mail is configured correctly

Thanks in advance.

### LINKED ISSUE #14013: No notifications are sent for mentions in review comments
- Gitea version (or commit ref): 1.13.0
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL)
  - [ ] No

## Description

When `@mentioning` someone in a code review comment, no notification is sent.

### LINKED ISSUE #14187: mentions are not persisted when ENABLE_NOTIFY_MAIL=false
Gitea version (or commit ref): 1.14.0

## Description

When ENABLE_NOTIFY_MAIL in app.ini is set to false, mentions in issues / pulls are not persisted (so not listed in the "Mentioning You" filter on the dashboard).

### PR DESCRIPTION
**Pull request by jpraet**
Fixes #14187: mention handling extracted from email notification code
Fixes #14013: add notification for mentions in pull request code comments
Fixes #13450: Not receiving any emails with setting "Only Email on Mention"

### CODE REVIEW

File: modules/notification/ui/ui.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func (ns *notificationService) NotifyCreateIssueComment(doer *models.User, repo
 	_ = ns.issueQueue.Push(opts)
 }
 
-func (ns *notificationService) NotifyNewIssue(issue *models.Issue) {
+func (ns *notificationService) NotifyNewIssue(issue *models.Issue, mentions []*models.User) {
 	_ = ns.issueQueue.Push(issueNotificationOpts{
```
Comment (lunny) [inline modules/notification/ui/ui.go:77]: Also we need send mentions notification to ui, of course that could be another PR.

**Comment by lunny** [comment]
May also resolved #13450

File: modules/notification/ui/ui.go (hunk 1 above)
Comment (jpraet) [reply modules/notification/ui/ui.go:77]: done

**Comment by jpraet** [comment]
> May also resolved #13450

I tested and it doesn't solve that problem. It is because the `visited map[int64]bool` is populated even when the notification wasn't sent. So when sending the mail to the mentioned users it thinks an email notification has already been sent. Easiest fix is probably to just switch the order around in `mailIssueCommentToParticipants`: first send to mentions, then to other participants?

I notice something else: you can explicitly unsubscribe from an issue to no longer receive notifications from it. However, in the case where someone explicitly mentions you, shouldn't we still send a notification?

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
`Unsubcribe` should not affect `mention`. But user has an option about mail notification. It could be `all`, `mentioned` and `none`.

**Comment by jpraet** [comment]
#13450 should be fixed now as well. And unsubscribe no longer affects mentions.

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@jpraet pleace send backport :)

**Pull request merged**